File: src/worldview_core/merge.py
"""Three-way merge of worldviews, for the fork-and-experiment workflow.

Given a common ancestor ``base`` and two descendants ``ours`` and
``theirs``, produce a merged worldview and a list of conflicts.  Entries
are matched by **local id** (the way a human tracks a statement across
edits of one lineage), and compared by **content**:

* statement content = canonical text, mode, meta, ext;
* argument content = premise set, conclusion set, canonical
  justification, rule, meta, ext.

Per id the usual rule applies: if both sides agree, take it; if only one
side changed (edited, added, or deleted) relative to base, take that
side; if both changed differently, it is a conflict and ``ours`` wins in
the merged output while the conflict is reported.  An argument left
referring to a statement that the other side deleted is a dangling
conflict and is dropped from the output.

Use :func:`diff` (identity-based) to *understand* how two worldviews
differ; use this to *combine* two lines of edits.
"""

from __future__ import annotations

from typing import Any

from .canon import canon
from .model import Argument, Statement, Worldview


def _stmt_key(s: Statement | None):
    if s is None:
        return None
    return (canon(s.text), s.mode, _freeze(s.meta), _freeze(s.ext))


def _arg_key(a: Argument | None):
    if a is None:
        return None
    return (frozenset(a.premises), frozenset(a.conclusions), canon(a.justification), a.rule, _freeze(a.meta), _freeze(a.ext))


def _freeze(x: Any):
    if isinstance(x, dict):
        return tuple(sorted((k, _freeze(v)) for k, v in x.items()))
    if isinstance(x, list):
        return tuple(_freeze(v) for v in x)
    return x


def _three_way(b, o, t, key):
    """Return (chosen, conflict) for one id."""
    kb, ko, kt = key(b), key(o), key(t)
    if ko == kt:
        return o, False
    if ko == kb:
        return t, False
    if kt == kb:
        return o, False
    return (o if o is not None else t), True


def _ordered_ids(base_ids, ours_ids, theirs_ids) -> list[str]:
    seen: set[str] = set()
    out: list[str] = []
    for seq in (base_ids, ours_ids, theirs_ids):
        for i in seq:
            if i not in seen:
                seen.add(i)
                out.append(i)
    return out


def merge(base: Worldview, ours: Worldview, theirs: Worldview) -> dict[str, Any]:
    conflicts: list[dict[str, Any]] = []
    summary = {"statements": {"kept": 0, "added_ours": 0, "added_theirs": 0, "added_both": 0, "removed": 0, "changed": 0},
               "arguments": {"kept": 0, "added_ours": 0, "added_theirs": 0, "added_both": 0, "removed": 0, "changed": 0}}

    def run(kind: str, b_items, o_items, t_items, key, describe):
        b = {x.id: x for x in b_items}
        o = {x.id: x for x in o_items}
        t = {x.id: x for x in t_items}
        result = []
        tally = summary[kind]
        for i in _ordered_ids(b, o, t):
            bi, oi, ti = b.get(i), o.get(i), t.get(i)
            chosen, conflict = _three_way(bi, oi, ti, key)
            if conflict:
                conflicts.append({
                    "kind": kind[:-1],  # "statement" / "argument"
                    "id": i,
                    "base": describe(bi),
                    "ours": describe(oi),
                    "theirs": describe(ti),
                    "resolution": "kept ours" if oi is not None else "kept theirs",
                })
            if chosen is None:
                if bi is not None:
                    tally["removed"] += 1
                continue
            result.append(chosen)
            if bi is None:
                if oi is not None and ti is not None:
                    tally["added_both"] += 1
                elif oi is not None:
                    tally["added_ours"] += 1
                else:
                    tally["added_theirs"] += 1
            elif key(chosen) == key(bi):
                tally["kept"] += 1
            else:
                tally["changed"] += 1
        return result

    statements = run("statements", base.statements, ours.statements, theirs.statements, _stmt_key, lambda s: s.to_dict() if s else None)
    arguments = run("arguments", base.arguments, ours.arguments, theirs.arguments, _arg_key, lambda a: a.to_dict() if a else None)

    # Dangling references: an argument survived but a statement it needs did not.
    present = {s.id for s in statements}
    kept_args = []
    for a in arguments:
        missing = [x for x in (*a.premises, *a.conclusions) if x not in present]
        if missing:
            conflicts.append({
                "kind": "dangling",
                "id": a.id,
                "missing": missing,
                "argument": a.to_dict(),
                "resolution": "dropped argument",
            })
        else:
            kept_args.append(a)

    # Header fields, same rule per field.
    def field(name):
        chosen, conflict = _three_way(getattr(base, name), getattr(ours, name), getattr(theirs, name), _freeze)
        if conflict:
            conflicts.append({"kind": "header", "id": name, "base": getattr(base, name), "ours": getattr(ours, name), "theirs": getattr(theirs, name), "resolution": "kept ours"})
        return chosen

    merged = Worldview(
        statements=statements,
        arguments=kept_args,
        name=field("name"),
        description=field("description"),
        version=ours.version,
        meta=field("meta"),
        ext=field("ext"),
    )
    return {"merged": merged.to_dict(), "conflicts": conflicts, "summary": summary}

File: sdk/typescript/src/merge.ts
/**
 * Three-way merge of worldviews, for the fork-and-experiment workflow.
 *
 * Given a common ancestor `base` and two descendants `ours` and `theirs`,
 * produce a merged worldview and a list of conflicts.  Entries are matched
 * by **local id** (the way a human tracks a statement across edits of one
 * lineage), and compared by **content**:
 *
 * - statement content = canonical text, mode, meta, ext;
 * - argument content = premise set, conclusion set, canonical
 *   justification, rule, meta, ext.
 *
 * Per id the usual rule applies: if both sides agree, take it; if only one
 * side changed (edited, added, or deleted) relative to base, take that
 * side; if both changed differently, it is a conflict and `ours` wins in
 * the merged output while the conflict is reported.  An argument left
 * referring to a statement that the other side deleted is a dangling
 * conflict and is dropped from the output.
 *
 * Use {@link diff} (identity-based) to *understand* how two worldviews
 * differ; use this to *combine* two lines of edits.
 *
 * Port of `merge.py`.  Content comparison reproduces Python's equality on
 * the JSON values of `meta` and `ext` (objects compared as sorted key
 * lists, `true == 1`, `1 == 1.0`), so the result is identical to the
 * reference implementation's.
 */

import { canon } from "./canon.js";
import { argumentToDict, statementToDict, worldviewToDict } from "./model.js";
import type { Argument, JsonObject, JsonValue, Statement, Worldview, WorldviewDocument } from "./model.js";

/** Counts per kind in {@link MergeReport.summary}. */
export interface MergeTally {
  kept: number;
  added_ours: number;
  added_theirs: number;
  added_both: number;
  removed: number;
  changed: number;
}

/** The header fields merged one by one. */
export type HeaderField = "name" | "description" | "meta" | "ext";

/** One conflict of {@link merge}. */
export type MergeConflict =
  | {
      kind: "statement";
      id: string;
      base: Statement | null;
      ours: Statement | null;
      theirs: Statement | null;
      resolution: "kept ours" | "kept theirs";
    }
  | {
      kind: "argument";
      id: string;
      base: Argument | null;
      ours: Argument | null;
      theirs: Argument | null;
      resolution: "kept ours" | "kept theirs";
    }
  | { kind: "dangling"; id: string; missing: string[]; argument: Argument; resolution: "dropped argument" }
  | {
      kind: "header";
      id: HeaderField;
      base: JsonValue | null;
      ours: JsonValue | null;
      theirs: JsonValue | null;
      resolution: "kept ours";
    };

/** The result of {@link merge}. */
export interface MergeReport {
  /** The merged worldview in file form: base order, then additions from ours, then from theirs. */
  merged: WorldviewDocument;
  conflicts: MergeConflict[];
  summary: { statements: MergeTally; arguments: MergeTally };
}

// -------------------------------------------------- Python-equal content keys

/** `merge.py`'s `_freeze`: a JSON value as nested tuples, dicts as sorted (key, value) pairs. */
type Frozen = null | string | number | boolean | Frozen[];

/** Python `str` ordering: by code point. */
function cpCompare(a: string, b: string): number {
  const ia = a[Symbol.iterator]();
  const ib = b[Symbol.iterator]();
  for (;;) {
    const x = ia.next();
    const y = ib.next();
    if (x.done) return y.done ? 0 : -1;
    if (y.done) return 1;
    const d = (x.value.codePointAt(0) as number) - (y.value.codePointAt(0) as number);
    if (d !== 0) return d;
  }
}

function freeze(x: JsonValue | undefined): Frozen {
  if (x === undefined || x === null) return null;
  if (Array.isArray(x)) return x.map(freeze);
  if (typeof x === "object") {
    return Object.keys(x)
      .sort(cpCompare)
      .map((k): Frozen => [k, freeze(x[k] as JsonValue)]);
  }
  return x;
}

/** Python `==` on frozen values: tuples elementwise, `True == 1 == 1.0`, everything else by type and value. */
function frozenEq(a: Frozen, b: Frozen): boolean {
  if (Array.isArray(a) || Array.isArray(b)) {
    if (!Array.isArray(a) || !Array.isArray(b) || a.length !== b.length) return false;
    return a.every((v, i) => frozenEq(v, b[i] as Frozen));
  }
  if (a === null || b === null) return a === b;
  if (typeof a === "string" || typeof b === "string") return a === b;
  return Number(a) === Number(b);
}

function stmtKey(s: Statement): Frozen {
  return [canon(s.text), s.mode, freeze(s.meta), freeze(s.ext)];
}

function idSet(ids: readonly string[]): Frozen {
  return [...new Set(ids)].sort();
}

function argKey(a: Argument): Frozen {
  return [idSet(a.premises), idSet(a.conclusions), canon(a.justification), a.rule ?? null, freeze(a.meta), freeze(a.ext)];
}

/** `_three_way`: the chosen value for one id and whether it was a conflict. */
function threeWay<T>(b: T | null, o: T | null, t: T | null, key: (x: T) => Frozen): [T | null, boolean] {
  const kb = b === null ? null : key(b);
  const ko = o === null ? null : key(o);
  const kt = t === null ? null : key(t);
  if (frozenEq(ko, kt)) return [o, false];
  if (frozenEq(ko, kb)) return [t, false];
  if (frozenEq(kt, kb)) return [o, false];
  return [o !== null ? o : t, true];
}

function byId<T extends { id: string }>(items: readonly T[]): Map<string, T> {
  const m = new Map<string, T>();
  for (const x of items) m.set(x.id, x);
  return m;
}

function orderedIds(...maps: Array<Map<string, unknown>>): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const m of maps) {
    for (const i of m.keys()) {
      if (!seen.has(i)) {
        seen.add(i);
        out.push(i);
      }
    }
  }
  return out;
}

function newTally(): MergeTally {
  return { kept: 0, added_ours: 0, added_theirs: 0, added_both: 0, removed: 0, changed: 0 };
}

// ---------------------------------------------------------------- merge

/** Three-way merge of two lines of edits from a common base.  See the module comment. */
export function merge(base: Worldview, ours: Worldview, theirs: Worldview): MergeReport {
  const conflicts: MergeConflict[] = [];
  const summary = { statements: newTally(), arguments: newTally() };

  function run<T extends { id: string }>(
    tally: MergeTally,
    bItems: readonly T[],
    oItems: readonly T[],
    tItems: readonly T[],
    key: (x: T) => Frozen,
    conflictOf: (id: string, b: T | null, o: T | null, t: T | null) => MergeConflict,
  ): T[] {
    const b = byId(bItems);
    const o = byId(oItems);
    const t = byId(tItems);
    const result: T[] = [];
    for (const i of orderedIds(b, o, t)) {
      const bi = b.get(i) ?? null;
      const oi = o.get(i) ?? null;
      const ti = t.get(i) ?? null;
      const [chosen, conflict] = threeWay(bi, oi, ti, key);
      if (conflict) conflicts.push(conflictOf(i, bi, oi, ti));
      if (chosen === null) {
        if (bi !== null) tally.removed++;
        continue;
      }
      result.push(chosen);
      if (bi === null) {
        if (oi !== null && ti !== null) tally.added_both++;
        else if (oi !== null) tally.added_ours++;
        else tally.added_theirs++;
      } else if (frozenEq(key(chosen), key(bi))) {
        tally.kept++;
      } else {
        tally.changed++;
      }
    }
    return result;
  }

  const statements = run(summary.statements, base.statements, ours.statements, theirs.statements, stmtKey, (id, b, o, t) => ({
    kind: "statement",
    id,
    base: b === null ? null : statementToDict(b),
    ours: o === null ? null : statementToDict(o),
    theirs: t === null ? null : statementToDict(t),
    resolution: o !== null ? "kept ours" : "kept theirs",
  }));
  const args = run(summary.arguments, base.arguments, ours.arguments, theirs.arguments, argKey, (id, b, o, t) => ({
    kind: "argument",
    id,
    base: b === null ? null : argumentToDict(b),
    ours: o === null ? null : argumentToDict(o),
    theirs: t === null ? null : argumentToDict(t),
    resolution: o !== null ? "kept ours" : "kept theirs",
  }));

  // Dangling references: an argument survived but a statement it needs did not.
  const present = new Set(statements.map((s) => s.id));
  const keptArgs: Argument[] = [];
  for (const a of args) {
    const missing = [...a.premises, ...a.conclusions].filter((x) => !present.has(x));
    if (missing.length > 0) {
      conflicts.push({ kind: "dangling", id: a.id, missing, argument: argumentToDict(a), resolution: "dropped argument" });
    } else {
      keptArgs.push(a);
    }
  }

  // Header fields, same rule per field.
  function field<K extends HeaderField>(name: K): Worldview[K] | undefined {
    type V = NonNullable<Worldview[K]>;
    const b = (base[name] ?? null) as V | null;
    const o = (ours[name] ?? null) as V | null;
    const t = (theirs[name] ?? null) as V | null;
    const [chosen, conflict] = threeWay<V>(b, o, t, (x) => freeze(x as JsonValue));
    if (conflict) {
      conflicts.push({ kind: "header", id: name, base: b as JsonValue | null, ours: o as JsonValue | null, theirs: t as JsonValue | null, resolution: "kept ours" });
    }
    return chosen === null ? undefined : chosen;
  }

  const merged: Worldview = { version: ours.version, statements, arguments: keptArgs };
  const name = field("name");
  if (name !== undefined) merged.name = name;
  const description = field("description");
  if (description !== undefined) merged.description = description;
  const meta = field("meta");
  if (meta !== undefined) merged.meta = meta as JsonObject;
  const ext = field("ext");
  if (ext !== undefined) merged.ext = ext as JsonObject;

  return { merged: worldviewToDict(merged), conflicts, summary };
}
